aws-cpp-sdk-core/source/utils/threading/Executor.cpp

```diff
@@ -107,3 +107,48 @@ bool PooledThreadExecutor::HasTasks()
     std::lock_guard<std::mutex> locker(m_queueLock);
     return m_tasks.size() > 0;
 }
+
+BlockingExecutor::BlockingExecutor(std::shared_ptr<Executor> executor, size_t poolSize) :
+m_executor(executor), m_poolSize(poolSize), m_numTasksRunning(0)
+{
+}
+
+BlockingExecutor::~BlockingExecutor()
+{
+}
+
+bool BlockingExecutor::SubmitToThread(std::function<void()>&& fn)
+{
+    std::unique_lock<std::mutex> locker(m_syncPointLock);
+    if (m_numTasksRunning < m_poolSize)
+    {
+        m_numTasksRunning++;
+    }
+    else
+    {
+        m_syncPoint.wait(locker);
+    }
+    locker.unlock();
+    
+    std::function<void()> executeTask = [&]() {

```

Comment:
This isn't free. A normal bind call would be cheaper. 
m_executor->Submit(&BlockingExecutor::ExecuteTask, this, fn);
this is the exact code we do in the async client calls.